### DISCUSSION TITLE
Invalidating infinite query leads to deleted pages from the cache

### DISCUSSION THREAD

**Discussion by eesamunir055** [TanStack/query#8456] [Q&A] [answered]
I have an infinite query setup like so for messages relating to a particular chat:

```
useInfiniteQuery({
        queryKey: ["messages", selectedChat?._id],
        queryFn: fetchMessages,
        initialPageParam: selectedChat?.latestMessage?.createdAt,
        getPreviousPageParam: (firstPage) => firstPage.previousTimestamp,
        getNextPageParam: (lastPage) => lastPage.nextTimestamp,
        staleTime: Infinity,
        enabled: !!selectedChat?._id,
        placeholderData: (prev) => prev,
    })

```
Within an realtime event handler, I receieve a delete event and call invalidate queries like so: `queryClient.invalidateQueries({queryKey: ["messages", selectedChat._id]})`. According to the docs, each page is refetched sequentially when invalidated, and this is what happens. The first page is refetched, however once the first page refetch completes, it removes all the other previously loaded pages in the cache, and then refetches them again in order.

Is there a way I can keep previously loaded pages in the cache until the refetch for these particular pages(caused by invalidateQueries) completes. I assumed `placeholderData` was the way to accomplish this however the same behaviour occurs. Is this expected, and if so is there anyway I can perhaps find a workaround?

**Comment by TkDodo** [comment]
> The first page is refetched, however once the first page refetch completes, it removes all the other previously loaded pages in the cache, and then refetches them again in order.

sounds like fetching the second page maybe fails? we had a bug around that, which was recently fixed, so make sure you’re on the latest version.

**Reply by eesamunir055** [reply]
After upgrading to the latest version, I noticed when invalidating, the issue of only one page appearing is fixed. If I initally load more or previous pages, everything works as expected and the `nextTimestamp` and `previousTimestamp` are as expected which tells me the logic for receiving the next/previous page is ok. However, after invalidating the query, it creates the same amount of pages however they are all duplicates of the first page that was in the cache before invalidation. All pages have the same data and timestamps/cursors. This new result (post-invalidation) doesn't include the other x amount of pages.

**Reply by TkDodo** [reply]
Would need to see a standalone reproduction please

**Reply by eesamunir055** [reply]
I understand. But thank you for helping me with  my  initial query

**Comment by eesamunir055** [comment, marked as answer]
Fixed this particular issue after updating, thanks TkDodo
